// protean/linked-list/item.js
//---------------------------------------------------------------------------
// List Item
//---------------------------------------------------------------------------
function Item (data, prev, next) {
    this.data = data;
    if (next) {
        next.prev = this;
    }
    this.next = next;
    if (prev) {
        prev.next = this;
    }
    this.prev = prev;
}

Item.prototype = {
    insert: function (item) {
        var n = this.next;
            
        this.next = item;
        item.next = n;
        item.prev = this;
        
        return this;
    },
    
    insertBefore: function (item) {
        var p = this.prev;
        
        this.prev = item;
        item.next = this;
        item.prev = p;
        
        return this;
    },
    
    destroy: function () {
        var p = this.prev,
            n = this.next;
        
        p.next = n;
        n.prev = p;
        
        this.data = this.prev = this.next = null;
        
        return this;
    },
    
    swap: function () {
        var n = this.next,
            p = this.prev;
        
        this.prev = n;
        if (n) {
            n.next = p;
        }
        
        this.next = p;
        if (p) {
            p.prev = n;
        }
        
        return this;
    }
};

module.exports = Item;
